### DISCUSSION TITLE
Use sentry

### DISCUSSION THREAD

**Discussion by rutonyru** [adonisjs/core#4599] [Help] [answered]
Hi, 
Well, after 3 months of dev, I'm still "emerveillé" (I love the french way to say "amazed") by your framework. 
As I heard that the good way to catch error is using sentry, I ask the question : is there a way to simply do that ? is it complicated ?
I didn't find a npm library : I tried lunifyfr/adonis-sentry (error with the installation) and octoper/adonis-sentry (unable to configure).
Thanks 
Antoine

**Comment by RomainLanz** [comment, marked as answer]
Hey @rutonyru! 👋🏻 

There is no need to use a dedicated library to use Sentry with AdonisJS.
You can directly install their SDK and start working from there.

Here is an example I have in production. It is a middleware that is executed to encapsulate the request and send telemetry information about it to Sentry (if you activate it inside Sentry).

```ts
import * as Sentry from '@sentry/node';
import type { HttpContext } from '@adonisjs/core/http';
import type { NextFn } from '@adonisjs/core/types/http';

export default class MonitoringMiddleware {
  async handle(ctx: HttpContext, next: NextFn) {
    const client = Sentry.getClient();
    const scope = new Sentry.Scope();
    scope.setClient(client);

    scope.setTag('url', ctx.request.url());

    ctx.sentry = scope;

    await Sentry.startSpan(
      {
        name: ctx.routeKey!,
        op: 'http.server',
        scope,
      },
      async () => {
        await next();
      },
    );
  }
}

declare module '@adonisjs/core/http' {
  export interface HttpContext {
    sentry: Sentry.Scope
  }
}
```

It also adds `sentry` inside your HttpContext, so you can use it to add information inside the request span.

Feel free to also add it inside the container binding.

```ts
ctx.containerResolver.bindValue(Sentry, sentry)
```

Allowing you to inject the Sentry instance inside any of your services.

Here is the provider that configures Sentry:
```ts
import * as Sentry from '@sentry/node';
import { nodeProfilingIntegration } from '@sentry/profiling-node';
import type { ApplicationService } from '@adonisjs/core/types';

export default class MonitoringProvider {
  constructor(protected app: ApplicationService) {}

  async boot() {
    if (this.app.inProduction) {
      const config = this.app.config.get<any>('sentry', {});

      Sentry.init({
        dsn: config.dsn,
        environment: config.environment,
        release: config.release,
        integrations: [nodeProfilingIntegration()],

        tracesSampleRate: 0.2,
        profilesSampleRate: 0.2,
      });
    }
  }
}
```

You can change the exception handler to add Sentry there too, so it can catch all your errors.

```ts
export default class HttpExceptionHandler extends ExceptionHandler {
  async report(error: unknown, ctx: HttpContext) {
    if (this.shouldReport(error) && app.inDev) {
      console.error(error)
      return
    }

    if (!this.shouldReport(error) || !app.inProduction) {
      return
    }

    ctx.sentry.setExtra('error', JSON.stringify(error))
    ctx.sentry.captureException(error)
    
    return super.report(error, ctx);
  }
}
```

**Reply by rutonyru** [reply]
Hi @RomainLanz, 
Nice answer. What a pleasure to be adonisjs addict :)
Have a nice day,

**Reply by RomainLanz** [reply]
I made that to simplify people story with Sentry inside an AdonisJS app: https://github.com/RomainLanz/sentry/
